File: ts/src/parser/statements/typedef.ts
import * as kw from "../keywords";
import { YangTypedefStmt } from "../../core/ast";
import { ParserContext, TokenStream, YangTokenType } from "../parser-context";
import { withMetadataSubstatements } from "../metadata-substatements";
import type { StatementParsers } from "../statement-parsers";

export class TypedefStatementParser {
  private readonly typedefBodyDispatch: Record<string, (tokens: TokenStream, context: ParserContext) => void>;

  constructor(private readonly parsers: StatementParsers) {
    this.typedefBodyDispatch = withMetadataSubstatements(this.parsers, {
      [kw.TYPE]: (tokens, context) => {
        this.parsers.type_parser.parse_type(tokens, context);
      },
      [kw.DEFAULT]: (tokens, context) => {
        this.parsers.parse_typedef_default(tokens, context);
      }
    });
  }

  parse_typedef(tokens: TokenStream, context: ParserContext): YangTypedefStmt {
    tokens.consume(kw.TYPEDEF);
    const name = tokens.consume_type(YangTokenType.IDENTIFIER);
    const stmt = new YangTypedefStmt({ name });
    const unsupportedCtx = `typedef '${name}'`;
    if (tokens.consume_if_type(YangTokenType.LBRACE)) {
      const child = context.push_parent(stmt);
      while (tokens.has_more() && tokens.peek_type() !== YangTokenType.RBRACE) {
        const handler = this.parsers.substatement_handler(tokens, this.typedefBodyDispatch);
        if (handler) {
          handler(tokens, child);
        } else if (!this.parsers.skip_unsupported_or_raise_unknown(tokens, unsupportedCtx)) {
          /* unreachable */
        }
      }
      tokens.consume_type(YangTokenType.RBRACE);
    }
    (context.module as any).typedefs[name] = stmt;
    // Keep nested typedefs on the parent (grouping/container/…) so ``uses``
    // expansion can re-register them on the importing module (RFC 7950 §7.13).
    this.parsers.add_to_parent_or_module(context, stmt);
    tokens.consume_if_type(YangTokenType.SEMICOLON);
    return stmt;
  }
}

File: src/xyang/parser/statements/typedef.py
"""
Parsing helpers for ``typedef`` statements.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Optional

from .. import keywords as kw

from ..metadata_substatements import with_metadata_substatements
from ..parser_context import TokenStream, ParserContext, YangTokenType
from ...ast import YangGroupingStmt, YangStatementList, YangTypedefStmt

if TYPE_CHECKING:
    from ..statement_parsers import StatementParsers


class TypedefStatementParser:
    """Parser for ``typedef`` statements."""

    def __init__(self, parsers: "StatementParsers") -> None:
        self._parsers = parsers
        self._typedef_body_dispatch = with_metadata_substatements(
            self._parsers,
            {
                kw.TYPE: self._parsers.parse_type,
                kw.DEFAULT: self._parsers.parse_typedef_default,
            },
        )

    def parse_typedef(
        self, tokens: TokenStream, context: ParserContext
    ) -> Optional[YangTypedefStmt]:
        """Parse typedef statement."""
        tokens.consume(kw.TYPEDEF)
        typedef_name = tokens.consume_type(YangTokenType.IDENTIFIER)
        typedef_stmt = YangTypedefStmt(name=typedef_name)
        unsupported_ctx = f"typedef '{typedef_name}'"

        if tokens.consume_if_type(YangTokenType.LBRACE):
            new_context = context.push_parent(typedef_stmt)
            while tokens.has_more() and tokens.peek_type() != YangTokenType.RBRACE:
                handler = self._parsers.substatement_handler(tokens, self._typedef_body_dispatch)
                if handler:
                    handler(tokens, new_context)
                elif self._parsers.skip_unsupported_or_raise_unknown_stmt(
                    tokens, unsupported_ctx
                ):
                    continue
            tokens.consume_type(YangTokenType.RBRACE)

        if typedef_name in context.module.typedefs:
            raise tokens.make_error(
                f"Duplicate typedef {typedef_name!r} in module {context.module.name!r}"
            )
        context.module.typedefs[typedef_name] = typedef_stmt
        parent = context.current_parent
        if isinstance(parent, YangGroupingStmt):
            parent.typedef_names.append(typedef_name)
        if isinstance(parent, YangStatementList) and parent is not context.module:
            parent.statements.append(typedef_stmt)
        tokens.consume_if_type(YangTokenType.SEMICOLON)
        return typedef_stmt
